### PR TITLE
Update signature input methods and requirements

### PR DESCRIPTION
**Pull request by jricher**
Update the signature input method to always require the signature parameters to be covered.

Redefine all fields like `@request-target` as specialty input fields (non-headers) and allow extension by registry.

﻿Define signature parameters using structured field semantics.

Define canonicalization rules more directly.

### CODE REVIEW

File: draft-ietf-httpbis-message-signatures.md [hunk 1]
```diff
@@ -744,6 +790,7 @@ The Signature Input is:
 ~~~
 *created: 1402170695
 *request-target: post /foo?param=value&pet=dog
+*signature-params: (*created *request-target); keyid="test-key-a"; created=1402170695
```
Comment (richanna) [inline draft-ietf-httpbis-message-signatures.md:793]:
```suggestion
*Signature-Input: mySignatureId=(*request-target); keyid="test-key-a"; created=1402170695
```

File: draft-ietf-httpbis-message-signatures.md [hunk 2]
```diff
@@ -323,6 +346,32 @@ Host: server.example.com</sourcecode></td>
...
+Note that an HTTP message could contain multiple signatures, but only the signature parameters used for the current signature are included.
+
+#### Canonicalization Examples
+
+Given the following signature parameters:
+
+|Property|Value|
+|--- |--- |
+|Algorithm|`hs2019`|
+|Covered Content|`@request-target`, `host`, `date`, `cache-contol`, `x-emptyheader`, `x-example`, `x-dictionary;key=b`, `x-dictionary;key=a`, `x-list;prefix=3`|
```
Comment (fimbault) [inline draft-ietf-httpbis-message-signatures.md:364]: small typo (cache-contol -> cache-control)

File: draft-ietf-httpbis-message-signatures.md [hunk 3]
```diff
@@ -161,21 +160,41 @@ Additionally, some examples use '\\' line wrapping for long values that contain
...
+Some content within HTTP messages can undergo transformations that change the bitwise value without altering meaning of the content (for example, the merging together of header fields with the same name).  Message content must therefore be canonicalized before it is signed, to ensure that a signature can be verified despite such intermediary transformations. This document defines rules for each content identifier that transform the identifier's associated content into such a canonical form.
 
-The following sections define content identifiers, their associated content, and their canonicalization rules.
+Content identifiers are defined using production grammar defined by {{!RFC8941}} section 4.
+The content identifier is an `sf-string` value. The content identifier
+type MAY define parameters which are included using the `parameters` rule.
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:169]: Where is the `parameters` rule defined?

File: draft-ietf-httpbis-message-signatures.md [hunk 4]
```diff
@@ -161,21 +160,41 @@ Additionally, some examples use '\\' line wrapping for long values that contain
...
-An HTTP header field value is canonicalized as follows:
+Unless overridden by additional parameters and rules, the HTTP header field value MUST be canonicalized with the following steps:
 
 1. Create an ordered list of the field values of each instance of the header field in the message, in the order that they occur (or will occur) in the message.
 2. Strip leading and trailing whitespace from each item in the list.
-3. Concatenate the list items together, with a comma "," and space " " between each item. The resulting string is the canonicalized value.
+3. Concatenate the list items together, with a comma "," and space " " between each item. 
+
+The resulting string is the canonicalized value.
+
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:190]: TODO: add example

File: draft-ietf-httpbis-message-signatures.md [hunk 5]
```diff
@@ -197,17 +216,19 @@ The following table shows example canonicalized values for header fields, given
...
+|`"server"`|www.example.com|
+|`"x-empty-header"`||
+|`"x-obs-fold-header"`|Obsolete line folding.|
+|`"x-ows-header"`|Leading and trailing whitespace.|
 {: title="Non-normative examples of header field canonicalization."}
 
 ## Dictionary Structured Field Members
 
-An individual member in the value of a Dictionary Structured Field is identified by the lowercased field name, followed by a semicolon `":"`, followed by the member name. An individual member in the value of a Dictionary Structured Field is canonicalized by applying the serialization algorithm described in Section 4.1.2 of {{!StructuredFields=I-D.ietf-httpbis-header-structure}} on a Dictionary containing only that member.
+An individual member in the value of a Dictionary Structured Field is identified by using the parameter `key` on the content identifier for the header. The value of this parameter is a the key being identified, without any parameters present on that key in the original dictionary.
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:229]: @jricher as discussed during IETF-109 this feature is still very problematic.

File: draft-ietf-httpbis-message-signatures.md [hunk 6]
```diff
@@ -244,46 +267,46 @@ The following table shows example canonicalized values for different content ide
...
+|`"x-list-a";prefix=6`|(a, b, c, d, e, f)|
+|`"x-list-b";prefix=0`|()|
 {: title="Non-normative examples of list prefix canonicalization."}
 
 
-## Signature Creation Time
+## Specialty Content Fields
 
-The signature's Creation Time ({{signature-metadata}}) is identified by the `*created` identifier.
+Content not found in an HTTP header can be included in the signature base string by defining a content identifier and the canonicalization method for its content.
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:280]: It is not clear whether content found in an HTTP header can be further included in the signature.

File: draft-ietf-httpbis-message-signatures.md [hunk 7]
```diff
@@ -244,46 +267,46 @@ The following table shows example canonicalized values for different content ide
...
 
-Its canonicalized value is an `sf-integer` containing the signature's Creation Time expressed in "Unix time".
+To differentiate speciality content identifiers from HTTP headers, specialty content identifiers MUST start with the "at" `@` character. This specification defines the following specialty content identifiers:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:282]: Use `specialty` for consistency

File: draft-ietf-httpbis-message-signatures.md [hunk 8]
```diff
@@ -333,27 +382,45 @@ HTTP Message Signatures have metadata properties that provide information regard
...
-: A timestamp representing the point in time that the signature was generated. Sub-second precision is not supported. A signature's Creation Time MAY be undefined, indicating that it is unknown.
-
-Covered Content:
-: An ordered list of content identifiers (Section 2) that indicates the metadata and message content that is covered by the signature. The order of identifiers in this list affects signature generation and verification, and therefore MUST be preserved.
+: A timestamp representing the point in time that the signature was generated, represented as an integer. Sub-second precision is not supported. A signature's Creation Time MAY be undefined, indicating that it is unknown. `sf-integer`
 
 Expiration Time:
-: A timestamp representing the point in time at which the signature expires. An expired signature always fails verification. A signature's Expiration Time MAY be undefined, indicating that the signature does not expire.
+: A timestamp representing the point in time at which the signature expires, represented as an integer. An expired signature always fails verification. A signature's Expiration Time MAY be undefined, indicating that the signature does not expire.
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:391]: signature always expire ;) better would be to say that it's unknown.

File: draft-ietf-httpbis-message-signatures.md (hunk 5 above)
Comment (jricher) [reply draft-ietf-httpbis-message-signatures.md:229]: I don't necessarily disagree but I wanted to at least make it consistent with the rest of the syntax. An upside is that we don't NEED this functionality in the same way that we used to (with the new `@signature-params` speciality field), and it's also more straightfoward to apply than it was before.

File: draft-ietf-httpbis-message-signatures.md (hunk 6 above)
Comment (jricher) [reply draft-ietf-httpbis-message-signatures.md:280]: Content found in an HTTP header can be included in the signature by adding the header to the signed content, which is described in its own section.

File: draft-ietf-httpbis-message-signatures.md (hunk 7 above)
Comment (jricher) [reply draft-ietf-httpbis-message-signatures.md:282]: "speciality" is a typo on my part that I didn't notice! It actually took me a few reads to find the problem here. :)

File: draft-ietf-httpbis-message-signatures.md (hunk 8 above)
Comment (jricher) [reply draft-ietf-httpbis-message-signatures.md:391]: I don't think that's actually the case. A signature may be perpetual depending on the use case. The amount of time that's passed might limit how much I'm willing to trust it, though. Especially for something like an HTTP message request!

File: draft-ietf-httpbis-message-signatures.md (hunk 3 above)
Comment (jricher) [reply draft-ietf-httpbis-message-signatures.md:169]: It's defined in RFC8491: https://www.rfc-editor.org/rfc/rfc8941.html#section-3.1.2-4

We need to have a more formal import of the ABNF rules in this document, but that hasn't been done yet.

File: draft-ietf-httpbis-message-signatures.md [hunk 9]
```diff
@@ -161,21 +160,41 @@ Additionally, some examples use '\\' line wrapping for long values that contain
...
+In order to allow signers and verifiers to establish which content is covered by a signature, this document defines content identifiers for data items covered by an HTTP Message Signature.
 
-Some content within HTTP messages may undergo transformations that change the bitwise value without altering meaning of the content (for example, the merging together of header fields with the same name).  Message content must therefore be canonicalized before it is signed, to ensure that a signature can be verified despite such innocuous transformations.  This document defines rules for each content identifier that transform the identifier's associated content into such a canonical form.
+Some content within HTTP messages can undergo transformations that change the bitwise value without altering meaning of the content (for example, the merging together of header fields with the same name).  Message content must therefore be canonicalized before it is signed, to ensure that a signature can be verified despite such intermediary transformations. This document defines rules for each content identifier that transform the identifier's associated content into such a canonical form.
```
[the remark this answers is not in the document]
Comment (jricher) [reply draft-ietf-httpbis-message-signatures.md:165]: Thanks, I missed that non-normative must when I made the other editorial changes here. 

The fact that some parties want to have evidence of transformation of some things is exactly why you specify what NEEDS to be signed when applying this specification. The start of that language was included in #1419

**Pull request merged**